### ISSUE TITLE
Incorrect warnings with deconstruction in `foreach`

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#35365]
```C#
using System.Collections.Generic;

static class Ext
{
    public static void Deconstruct<TKey, TValue>(
        this KeyValuePair<TKey, TValue> pair, out TKey key, out TValue value)
    {
        key = pair.Key;
        value = pair.Value;
    }
}

class Program
{
    static void F(IEnumerable<KeyValuePair<string, int>> pairs)
    {
        foreach (var (k, v) in pairs)
        {
        }
    }
}
```
```
(17,18): warning CS8619: Nullability of reference types in value of type
    'KeyValuePair<string, int>' doesn't match target type 'string'.
(17,18): warning CS8619: Nullability of reference types in value of type
    'KeyValuePair<string, int>' doesn't match target type 'int'.
```

**Comment by cston** [comment]
Duplicate of https://github.com/dotnet/roslyn/issues/35131.
